### ISSUE TITLE
Should schema.Schema, Fields and FieldValidators be splitt into a compiled and uncompiled type?

### ISSUE THREAD

**Issue by smyrman** [rs/rest-layer#194]
A discussion in #109 suggest distinguishing between a `resource.IndexBuilder` (allows adding resources, and returning a compiled index) and `resource.Index` (allows finding active resources and fetching items).

There are some reasons that could make it worthwhile to take a similar approach for schema.Schema:
- FieldValidors are often referred to as pointers (e.g. `schema.IDField`, to mention one). When the Field definition is reused in multiple schemas, or the same schema is compiled multiple times in parallel, e.g. for tests, a data race condition arise.  When CIs run the tests with race detection enabled, they are likely to fail.
- We could avoid doing things like #110 - instead, `.Compile` would be a method on a `FieldValidatorBuilder` returning a `FieldValidator` (naming not final).
- We won't be able to _accidentally_ use an uncompiled schema in place of a compiled one.

This changes should be considered alongside other potentially large schema package changes, such as #77, that basically suggests allowing _any_ FieldValidator to be used a the Schema top-level.

**Comment by smyrman** [comment]
PS! Not something I am going to work on anytime soon.

**Comment by rs** [comment]
Interesting idea.

**Comment by smyrman** [comment]
Plausible interface.

To illustrate with today's naming (i.e. without #77):

```go
type Index interface {
    GetResource(name string) query.Resource
}

type Type interface{
    Validator(field Field, idx Index) FieldValidator
    Serializer(field Field, idx Index) FieldSerializer
}
```
Where the new `Index` interface may be used for other things as well, such as Reference checks (replace `RefrenceChecker`) and embedding (see #129).

If  assuming `schema.Field` and `schema.Schema` is "merged" as suggested in #77, the code becomes:

```go
type Type interface{
    Validator(schema Schema, idx Index) Validator
    Serializer(schema Schema, idx Index) Serializer
}
```
I am picturing that these calls a re _recursive_ so that when you _Bind_ a new resource against a `resource.Index` with a given `Schema, you extract the Validator and Serializer, and potentially trow away the schema (unless it's needed for something else, like documentation generation).

**Comment by smyrman** [comment]
Will close issues that relates to a broader redesign of the schema package (such as this one), leaving only #77 open.
